File: lib/services/content_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/comment.dart';
import '../models/profile_models.dart';

/// Serviço para gerenciar conteúdo e interações via API
class ContentService {
  final String baseUrl;

  ContentService({required this.baseUrl});

  /// Busca comentários para um conteúdo específico
  /// @param contentId ID do conteúdo
  /// @returns Lista de comentários
  Future<List<Comment>> getCommentsForContent(String contentId) async {
    final response = await http.get(Uri.parse('$baseUrl/content/$contentId/comments'));

    if (response.statusCode == 200) {
      final List data = jsonDecode(response.body);
      return data.map((item) => Comment.fromJson(item)).toList();
    } else {
      throw Exception('Erro ao buscar comentários');
    }
  }

  /// Adiciona um comentário a um conteúdo
  /// @param contentId ID do conteúdo
  /// @param userId ID do usuário
  /// @param text Texto do comentário
  /// @returns Comentário criado
  Future<Comment> addComment(String contentId, String userId, String text) async {
    final response = await http.post(
      Uri.parse('$baseUrl/content/$contentId/comments'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'userId': userId,
        'text': text,
      }),
    );

    if (response.statusCode == 201) {
      final data = jsonDecode(response.body);
      return Comment.fromJson(data);
    } else {
      throw Exception('Erro ao adicionar comentário');
    }
  }

  /// Busca vídeos similares baseado em categorias
  /// @param contentId ID do conteúdo atual
  /// @param categories Lista de categorias
  /// @returns Lista de vídeos similares
  Future<List<ProfileContent>> getSimilarVideos(String contentId, List<String> categories) async {
    final response = await http.post(
      Uri.parse('$baseUrl/content/$contentId/similar'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'categories': categories}),
    );

    if (response.statusCode == 200) {
      final List data = jsonDecode(response.body);
      return data.map((item) => ProfileContent.fromJson(item)).toList();
    } else {
      throw Exception('Erro ao buscar vídeos similares');
    }
  }

  /// Busca recomendações para um conteúdo
  /// @param contentId ID do conteúdo
  /// @returns Lista de conteúdos recomendados
  Future<List<ProfileContent>> getRecommendations(String contentId) async {
    final response = await http.get(Uri.parse('$baseUrl/content/$contentId/recommendations'));

    if (response.statusCode == 200) {
      final List data = jsonDecode(response.body);
      return data.map((item) => ProfileContent.fromJson(item)).toList();
    } else {
      throw Exception('Erro ao buscar recomendações');
    }
  }

  /// Incrementa visualizações de um conteúdo
  /// @param contentId ID do conteúdo
  Future<void> incrementViews(String contentId) async {
    final response = await http.post(Uri.parse('$baseUrl/content/$contentId/views'));

    if (response.statusCode != 200) {
      throw Exception('Erro ao incrementar visualizações');
    }
  }

  /// Cria novo conteúdo
  /// @param userId ID do usuário criador
  /// @param contentData Dados do conteúdo
  /// @returns Conteúdo criado
  Future<ProfileContent> createContent(String userId, Map<String, dynamic> contentData) async {
    final response = await http.post(
      Uri.parse('$baseUrl/content/$userId'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(contentData),
    );

    if (response.statusCode == 201) {
      final data = jsonDecode(response.body);
      return ProfileContent.fromJson(data);
    } else {
      throw Exception('Erro ao criar conteúdo');
    }
  }
}

File: backend/services/contentService.js
const db = require('../config/db');

exports.getProfileContent = async (userId, type, limit = 10) => {
  console.log('[ContentService] Buscando conteúdo do perfil:', userId, type);

  try {
    let query = 'SELECT * FROM content WHERE user_id = ?';
    let params = [userId];

    if (type !== 'all') {
      // Map plural types to singular
      let dbType = type;
      if (type === 'posts') dbType = 'post';
      else if (type === 'videos') dbType = 'video';
      else if (type === 'exclusive') dbType = 'exclusive';

      query += ' AND type = ?';
      params.push(dbType);
    }

    query += ' ORDER BY created_at DESC LIMIT ?';
    params.push(limit);

    const [rows] = await db.query(query, params);

    return rows.map(row => {
      return {
        id: row.id.toString(),
        title: row.title,
        type: row.type,
        thumbnailUrl: row.thumbnail_url,
        videoUrl: row.video_url,
        category: Array.isArray(row.category) ? row.category : [],
        tags: Array.isArray(row.tags) ? row.tags : [],
        duration: row.duration,
        views: row.views,
        createdAt: row.created_at
      };
    });
  } catch (err) {
    console.error('[ContentService] Erro ao buscar conteúdo:', err);
    throw err;
  }
};

exports.createContent = async (userId, contentData) => {
  console.log('[ContentService] Criando conteúdo:', contentData.title);

  try {
    const [result] = await db.query(
      'INSERT INTO content (user_id, title, type, thumbnail_url, video_url, category, tags, duration, tier_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)',
      [
        userId,
        contentData.title,
        contentData.type,
        contentData.thumbnailUrl,
        contentData.videoUrl,
        JSON.stringify(contentData.category || []),
        JSON.stringify(contentData.tags || []),
        contentData.duration,
        contentData.tierId
      ]
    );

    // Atualizar contador de posts no perfil
    await db.query('UPDATE profile_stats SET posts = posts + 1 WHERE user_id = ?', [userId]);

    return {
      id: result.insertId.toString(),
      ...contentData,
      createdAt: new Date()
    };
  } catch (err) {
    console.error('[ContentService] Erro ao criar conteúdo:', err);
    throw err;
  }
};

exports.getCommentsForContent = async (contentId) => {
  console.log('[ContentService] Buscando comentários:', contentId);

  try {
    const [rows] = await db.query(`
      SELECT c.*, u.name as user_name, u.avatar_url
      FROM comments c
      JOIN users u ON c.user_id = u.id
      WHERE c.content_id = ?
      ORDER BY c.created_at DESC
    `, [contentId]);

    return rows.map(row => ({
      id: row.id.toString(),
      contentId: row.content_id.toString(),
      userId: row.user_id.toString(),
      text: row.text,
      likes: row.likes,
      createdAt: row.created_at,
      userName: row.user_name,
      userAvatar: row.avatar_url
    }));
  } catch (err) {
    console.error('[ContentService] Erro ao buscar comentários:', err);
    throw err;
  }
};

exports.addComment = async (contentId, userId, text) => {
  console.log('[ContentService] Adicionando comentário:', contentId);

  try {
    const [result] = await db.query(
      'INSERT INTO comments (content_id, user_id, text) VALUES (?, ?, ?)',
      [contentId, userId, text]
    );

    return {
      id: result.insertId.toString(),
      contentId: contentId.toString(),
      userId: userId.toString(),
      text,
      likes: 0,
      createdAt: new Date()
    };
  } catch (err) {
    console.error('[ContentService] Erro ao adicionar comentário:', err);
    throw err;
  }
};

exports.getSimilarVideos = async (contentId, categories) => {
  console.log('[ContentService] Buscando vídeos similares:', contentId);

  try {
    // Buscar vídeos com categorias similares
    const [rows] = await db.query(`
      SELECT * FROM content
      WHERE type = 'video' AND id != ?
      AND JSON_OVERLAPS(category, ?)
      ORDER BY views DESC
      LIMIT 5
    `, [contentId, JSON.stringify(categories)]);

    return rows.map(row => ({
      id: row.id.toString(),
      title: row.title,
      type: row.type,
      thumbnailUrl: row.thumbnail_url,
      videoUrl: row.video_url,
      category: row.category ? JSON.parse(row.category) : [],
      tags: row.tags ? JSON.parse(row.tags) : [],
      duration: row.duration,
      views: row.views,
      createdAt: row.created_at
    }));
  } catch (err) {
    console.error('[ContentService] Erro ao buscar vídeos similares:', err);
    throw err;
  }
};

exports.getRecommendations = async (contentId) => {
  console.log('[ContentService] Buscando recomendações:', contentId);

  try {
    const [rows] = await db.query('SELECT * FROM recommendations WHERE content_id = ?', [contentId]);
    if (rows.length === 0) return [];

    const recommendedIds = JSON.parse(rows[0].recommended_content_ids);
    if (recommendedIds.length === 0) return [];

    const [contentRows] = await db.query(`SELECT * FROM content WHERE id IN (${recommendedIds.map(() => '?').join(',')})`, recommendedIds);

    return contentRows.map(row => ({
      id: row.id.toString(),
      title: row.title,
      type: row.type,
      thumbnailUrl: row.thumbnail_url,
      videoUrl: row.video_url,
      category: row.category ? JSON.parse(row.category) : [],
      tags: row.tags ? JSON.parse(row.tags) : [],
      duration: row.duration,
      views: row.views,
      createdAt: row.created_at
    }));
  } catch (err) {
    console.error('[ContentService] Erro ao buscar recomendações:', err);
    throw err;
  }
};

exports.incrementViews = async (contentId) => {
  console.log('[ContentService] Incrementando visualizações:', contentId);

  try {
    await db.query('UPDATE content SET views = views + 1 WHERE id = ?', [contentId]);
    return true;
  } catch (err) {
    console.error('[ContentService] Erro ao incrementar visualizações:', err);
    throw err;
  }
};
